crates/subspace-service/src/piece_cache.rs

```diff
@@ -150,15 +154,21 @@ where
                 .collect::<Vec<_>>(),
         )?;
 
-        for piece_index in delete_indexes {
-            self.local_provided_keys.remove(&piece_index);
-        }
+        let local_provided_keys = {
```

Comment:
Looking at this again, I feel like we're over-engineering a lot here. All we need to store is the first `piece_index`, the rest can be derived from it using `max_pieces_in_cache`. Doesn't have to be done in this PR though.